### DISCUSSION TITLE
[CS2] Performance tests on Hybrid GPU configurations (laptops)

### DISCUSSION THREAD

**Discussion by tcharchar** [Blinue/Magpie#925] [Show and tell]
### Program(Game) 游戏名

Counter-Strike 2

### Capture method 捕获模式

All

### Other information (optional) 补充信息（可选）

> [!NOTE]
> While I am aware this tool is not really meant for intensive modern games, I use it primarily for FPS/TPS games in order to play in stretched resolutions without changing my own display's resolution.
>
> GDI and DwmSharedSurface show a pure black screen regardless of configuration, this is probably intended.
>
> All of these obviously have a input delay higher than proper exclusive fullscreen, but it could be better if Magpie could run over refresh rate, to minimize frametime delay as much as possible.

This is a compilation of various tests and their outcomes. 
All use nearest neighbor set on fill to prevent any post-processing delays as much as possible. 

Tests are all on a laptop (NVIDIA RTX 4070 Mobile + Intel Iris Xe (i7-13700H))
CS2 runs in 1024x768 windowed, with my display's actual resolution being 2560x1600@165
CS2 runs at 280-320 FPS without the use of Magpie.
All tests were made with CS2 using the NVIDIA GPU. Optimus/NVIDIA only was set with the help of NVIDIA's control panel.

## Optimus 
| GPU used in Magpie | Capture Method | Behaviour | 
| :-: | ------------- | ------------- | 
| Intel | Graphics Capture | CS2 runs at ~90fps with inconsistent drops to 40fps. Magpie runs at 140-165 FPS. Frametimes are horrible. |
| Intel | Desktop Duplication | CS2 runs at 90-120fps. Magpie's FPS attempts catching up to the game (or slows down the game), causing stutters to be longer than they should. |
| NVIDIA | Graphics Capture | CS2 runs at ~180-220FPS. Magpie runs at 165FPS. Frametimes aren't great, but very much playable. |
| NVIDIA | Desktop Duplication | Magpie refuses to scale the game. This is (probably) expected behaviour. |

## NVIDIA GPU only
| GPU used in Magpie | Capture Method | Behaviour | 
| :-: | ------------- | ------------- | 
| Intel | Graphics Capture | CS2 runs at ~190-250FPS. Magpie runs at 165FPS. Overall recommended. |
| Intel | Desktop Duplication | CS2 runs at ~180-220FPS. Magpie runs at 120-145FPS. Frametimes are great, but the random FPS drops cause it to be unfavourable. |
| NVIDIA | Graphics Capture | CS2 runs at ~110fps with constant drops to 70fps. Magpie runs at 140-165 FPS. Frametimes are worse than Optimus+Intel+Graphics Capture. |
| NVIDIA | Desktop Duplication | Magpie refuses to scale the game. This is unexpected behaviour, as the Intel GPU is no longer in charge of rendering the display. |

Even though this isn't probably helpful to any fixes in order to alleviate the issues in the other rendering modes, I still hope this helps whoever wants a quick "best settings" for Magpie and Shooter games with multiple GPUs.

### Log files 日志文件

[magpie.log](https://github.com/Blinue/Magpie/files/15410161/magpie.log)

**Comment by Blinue** [comment]
Thank you for your testing!

Since Magpie is a GPU-intensive program, I do not recommend using it for 3D games. Many people claim that Magpie can improve frame rates, but this is not a supported use case. Whether frame rates improve depends heavily on hardware and drivers, so please prioritize the game's built-in super-resolution feature.

I'll convert this issue to a discussion.

**Reply by tcharchar** [reply]
Thank you for converting this to the proper format.
While not perfect nor its desired use case, I have found it to be really useful due to how low I usually drop my resolutions in shooter games (CS2, Fortnite, TF2, Apex and others whom I can't think of) while using a nice method to resample/upscale (and stretch) the image without a display's rescaling blurry artifacts, or other various issues (unsupported custom resolution either by the game or the display, desktop icon repositioning, screen flickering on game launch due to resolution change etc.).

Also should've mentioned, I'm not particularly seeking better FPS. I'm willing to trade a chunk of performance for on-the-fly app/game stretching 😁.
